packages/tangle/booker.go

```diff
@@ -65,24 +64,56 @@ func (b *Booker) Setup() {
 }
 
 // BookMessage tries to book the given Message (and potentially its contained Transaction) into the LedgerState and the Tangle.
-// It fires a MessageBooked event if it succeeds.
+// It fires a MessageBooked event if it succeeds. If the Message is invalid it fires a MessageInvalid event.
 func (b *Booker) BookMessage(messageID MessageID) (err error) {
 	b.tangle.Storage.Message(messageID).Consume(func(message *Message) {
 		b.tangle.Storage.MessageMetadata(messageID).Consume(func(messageMetadata *MessageMetadata) {
-			// don't book the same message more than once!
+			// Don't book the same message more than once!
 			if messageMetadata.IsBooked() {
 				err = errors.Errorf("message already booked %s", messageID)
 				return
 			}
 
+			// Like references need to point to messages containing transactions
+			if !b.allMessagesContainTransactions(message.ParentsByType(LikeParentType)) {
+				messageMetadata.SetInvalid(true)
+				b.tangle.Events.MessageInvalid.Trigger(messageID)
+				err = errors.Errorf("message like reference does not contain a transaction %s", messageID)
+				return
+			}
+
+			// We obtain strong parents branches using metadata and past markers
+			strongBranchIDs := b.strongParentsBranchIDs(message)
+			// We obtain liked payload branches
+			likedBranchIDs := b.likedParentsBranchIDs(message)
+
+			// We collect strong parents branches recursively
+			prunedCollectedStrongParents := b.collectBranchesUpwards(strongBranchIDs)
+			// For every liked branch we need to prune it and all its descendants from the collected strong branches
+			for likedBranchID := range likedBranchIDs {
+				b.tangle.LedgerState.BranchDAG.ForEachConflictingBranchID(likedBranchID, func(conflictingBranchID ledgerstate.BranchID) {
+					// If we like something in the same conflict set of a collected strong parent
+					// we remove the collected strong parent from the collection, including all its children
+					prunedCollectedStrongParents.Subtract(b.collectBranchesDownwards(conflictingBranchID))
+				})
+			}
+
+			// We filter our strong parents and add the liked parents to the resulting set
+			supportedBranches := strongBranchIDs.Intersect(prunedCollectedStrongParents).AddAll(likedBranchIDs)
```

Comment:
Should be refactored out into a separate function; something like `SupportedBranches()` or whatever with a doc comment explaining how these are determined.